RavenDB-15601 Additional option to collect debug-info

YonatanResh opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-15601/Additional-option-to-collect-debug-info

### Additional description

in the UI we need to add the option to choose if you want to get the server wide, log file, and all the databases or a subset of the databases

### Type of change

- New feature

### How risky is the change?

- Low 

### Backward compatibility

- Non breaking change

### Testing by Contributor

- It has been verified by manual testing

### UI work

- It requires further work in the Studio

YonatanResh committed:
RavenDB-15601 Additional option to collect debug-info

YonatanResh committed:
RavenDB-15601 Additional option to collect debug-info

ppekrol commented on src/Raven.Server/Documents/Handlers/Debugging/ServerWideDebugInfoPackageHandler.cs:
```diff
@@ -227,9 +255,17 @@ await using (var sw = new StreamWriter(entryStream))
+                url += $"&type={type}";
+
+            if (string.IsNullOrEmpty(databases) == false)
+                url += $"&databases=${string.Join("&databases=", databases)}";
```
name needs to be encoded

YonatanResh replied:
done

ppekrol commented on src/Raven.Server/Documents/Handlers/Debugging/ServerWideDebugInfoPackageHandler.cs:
```diff
@@ -449,5 +497,15 @@ internal class NodeDebugInfoRequestHeader
         }
+
+        [Flags]
+        public enum DebugInfoContents
```
DebugInfoPackageContent

YonatanResh replied:
done

ppekrol commented on src/Raven.Server/Documents/Handlers/Debugging/ServerWideDebugInfoPackageHandler.cs:
```diff
@@ -292,6 +328,18 @@ private async Task WriteForAllLocalDatabases(ZipArchive archive, JsonOperationCo
+                {
+                    var allDatabases = ServerStore.Cluster.GetDatabaseNames(context);
+
+                    if (string.IsNullOrEmpty(databases) == false && databases.Any())
```
this should be `databases.Count > 0`, it is a struct

YonatanResh replied:
done

ppekrol commented on src/Raven.Server/Documents/Handlers/Debugging/ServerWideDebugInfoPackageHandler.cs:
```diff
@@ -167,6 +170,8 @@ public async Task GetInfoPackage()
             HttpContext.Response.Headers["Content-Type"] = "application/zip";
 
+            var debugInfoType = GetDebugInfoType();
+            var databases = GetStringValuesQueryString("databases", required: false);
```
by convention, this should be `database` not `databases`

you are passing via url "database=X&database=Y" - passing 'databases' does not look natural here

YonatanResh replied:
done

ppekrol commented on src/Raven.Server/Documents/Handlers/Debugging/ServerWideDebugInfoPackageHandler.cs:
```diff
@@ -108,8 +109,10 @@ public async Task GetClusterWideInfoPackage()
 
             var timeoutInSecPerNode = GetIntValueQueryString("timeoutInSecPerNode", false) ?? 60;
             var clusterOperationToken = CreateOperationToken();
+            var type = GetStringQueryString("type", false);
```
use named paramter

YonatanResh replied:
done

YonatanResh committed:
RavenDB-15601 addressed PR comments

ppekrol commented on src/Raven.Server/Documents/Handlers/Debugging/ServerWideDebugInfoPackageHandler.cs:
```diff
@@ -227,9 +256,16 @@ await using (var sw = new StreamWriter(entryStream))
+                url += $"&type={type}";
+
+            if (databases.Count > 0)
+                url += $"&database={WebUtility.UrlEncode(string.Join("&database=", databases))}";
```
```
Uri.EscapeDataString(_indexName)
```

YonatanResh replied:
done

YonatanResh committed:
RavenDB-15601 addressed PR comments

ppekrol committed:
RavenDB-15601
- error early when parsing DebugInfoPackageContentType and it is invalid
- fixed order of arguments to GetDebugInfoFromNodeAsync
